canonical-web-and-design/maas-ui
# refactor(ui): improve action form composability

## Done

- Replaced active/selected machine selectors with a `machines` prop for each machine action form. This should make it easier to migrate forms to be able to work across node types, and also meant I could remove a ton of duplicated tests used for handling the active/selected action dispatches.
- Moved a bunch of logic out of the individual forms and into the form wrapper (`ActionFormWrapper`) so it all happens centrally. This reduces a ton of duplicated code.
- Removed some iffy workarounds in `ActionForm` to get the processing stuff to work.
- Removed `useMachineActionForm` hook since this is now handled exclusively in the machine action form wrapper
- Updated `useProcessing` hook to return completed state

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- I've set up a bunch of Manual machines in bolla so it might be easier to use bolla here
- Go to the machine list, select all the machines then select an action that can't be performed by all of them
- Check that a warning shows and you can filter the machines to only those that can perform the action
- Unselect the machines and select a bunch of "New" machines
- Commission them and check that the warning doesn't show up at any point
- With the same machines selected, mark them broken while they're commissioning
- Now, try and mark them fixed
- Check that an error shows and the form does not close
- Go to the details page of one of the broken machines and delete it
- Check that you are redirected to the machine list
- Check that you can still perform actions on VMs in the KVM details page

## Fixes
